Use lower lmdb memory allocation in tests to avoid failure

> Thanks for opening this issue.

:+1: 

> What evidence is there that the current solution isn't enough? 

When running groups of tests that include multiple concurrent running ursulas (like the integration suite) an exception is raised:

```
    @pytest.fixture(scope="module")
    def federated_ursulas(ursula_federated_test_config):
>       _ursulas = make_federated_ursulas(ursula_config=ursula_federated_test_config,
                                          quantity=NUMBER_OF_URSULAS_IN_DEVELOPMENT_NETWORK)

../fixtures.py:367: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../utils/ursula.py:69: in make_federated_ursulas
    ursula = ursula_config.produce(rest_port=port + 100,
../../nucypher/config/characters.py:109: in produce
    ursula = self.CHARACTER_CLASS(**merged_parameters)
../../nucypher/characters/lawful.py:1022: in __init__
    rest_app, datastore = make_rest_app(
../../nucypher/network/server.py:92: in make_rest_app
    datastore = datastore.Datastore(db_filepath)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <nucypher.datastore.datastore.Datastore object at 0x2c69d5fbe7c0>
db_path = '/tmp/tmpklq6py8i'

    def __init__(self, db_path: str) -> None:
        """
        Initializes a Datastore object by path.
    
        :param db_path: Filepath to a lmdb database.
        """
        self.db_path = db_path
>       self.__db_env = lmdb.open(db_path, map_size=self.LMDB_MAP_SIZE)
E       lmdb.MemoryError: /tmp/tmpklq6py8i: Cannot allocate memory

../../nucypher/datastore/datastore.py:96: MemoryError

```

> Is the problem still apparent after machine restart or equivalent?

Yes - on my development laptop a full hardware restart with minimal applications running still causes thus issue to raise an exception.

--- a/nucypher/characters/lawful.py
+++ b/nucypher/characters/lawful.py
@@ -41,7 +41,12 @@ import nucypher
 from bytestring_splitter import BytestringKwargifier, BytestringSplitter, BytestringSplittingError, \
     VariableLengthBytestring
 from constant_sorrow import constants
-from constant_sorrow.constants import INCLUDED_IN_BYTESTRING, PUBLIC_ONLY, STRANGER_ALICE, UNKNOWN_VERSION, READY
+from constant_sorrow.constants import (INCLUDED_IN_BYTESTRING,
+                                       PUBLIC_ONLY,
+                                       STRANGER_ALICE,
+                                       UNKNOWN_VERSION,
+                                       READY,
+                                       INVALIDATED)
 from nucypher.acumen.nicknames import nickname_from_seed
 from nucypher.acumen.perception import FleetSensor
 from nucypher.blockchain.eth.actors import BlockchainPolicyAuthor, Worker
@@ -1278,6 +1283,20 @@ class Ursula(Teacher, Character, Worker):
                 self._arrangement_pruning_task.stop()
         if halt_reactor:
             reactor.stop()
+
+    def _finalize(self):
+        """
+        Cleans up Ursula from objects that may eat up system resources.
+        Useful for testing purposes, where many Ursulas are created and destroyed,
+        and references to them may persist for too long.
+        This method is not needed if all references to the Ursula are released.
+
+        **Warning:** invalidates the Ursula.
+        """
+
+        # `rest_server` holds references to the datastore (directly and via `rest_app`).
+        # An open datastore hogs up file descriptors.
+        self.rest_server = INVALIDATED
 
     def rest_information(self):
         hosting_power = self._crypto_power.power_ups(TLSHostingPower)

--- a/nucypher/network/server.py
+++ b/nucypher/network/server.py
@@ -18,13 +18,14 @@ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
 import binascii
 import os
 import uuid
+import weakref
 from bytestring_splitter import BytestringSplitter
 from constant_sorrow import constants
 from constant_sorrow.constants import FLEET_STATES_MATCH, NO_BLOCKCHAIN_CONNECTION, NO_KNOWN_NODES
 from flask import Flask, Response, jsonify, request
 from hendrix.experience import crosstown_traffic
 from jinja2 import Template, TemplateError
-from typing import Tuple
+from typing import Tuple, Set
 from umbral.keys import UmbralPublicKey
 from umbral.kfrags import KFrag
 from web3.exceptions import TimeExhausted
@@ -36,7 +37,7 @@ from nucypher.crypto.kits import UmbralMessageKit
 from nucypher.crypto.powers import KeyPairBasedPower, PowerUpError
 from nucypher.crypto.signing import InvalidSignature
 from nucypher.crypto.utils import canonical_address_from_umbral_key
-from nucypher.datastore.datastore import RecordNotFound, DatastoreTransactionError
+from nucypher.datastore.datastore import Datastore, RecordNotFound, DatastoreTransactionError
 from nucypher.datastore.keypairs import HostingKeypair
 from nucypher.datastore.models import PolicyArrangement, Workorder
 from nucypher.network import LEARNING_LOOP_VERSION
@@ -80,16 +81,31 @@ class ProxyRESTServer:
 def make_rest_app(
         db_filepath: str,
         this_node,
-        serving_domains,
-        log=Logger("http-application-layer")
-        ) -> Tuple:
+        serving_domains: Set[str],
+        log: Logger=Logger("http-application-layer")
+        ) -> Tuple[Flask, Datastore]:
+    """
+    Creates a REST application and an associated ``Datastore`` object.
+    Note that the REST app **does not** hold a reference to the datastore;
+    it is your responsibility to ensure it lives for as long as the app does.
+    """
 
-    forgetful_node_storage = ForgetfulNodeStorage(federated_only=this_node.federated_only)
-
-    from nucypher.datastore import datastore
+    # A trampoline function for the real REST app,
+    # to ensure that a reference to the node and the datastore object is not held by the app closure.
+    # One would think that it's enough to only remove a reference to the node,
+    # but `rest_app` somehow holds a reference to itself, Uroboros-like,
+    # and will hold the datastore reference if it is created there.
 
     log.info("Starting datastore {}".format(db_filepath))
-    datastore = datastore.Datastore(db_filepath)
+    datastore = Datastore(db_filepath)
+    rest_app = _make_rest_app(weakref.proxy(datastore), weakref.proxy(this_node), serving_domains, log)
+
+    return rest_app, datastore
+
+
+def _make_rest_app(datastore: Datastore, this_node, serving_domains: Set[str], log: Logger) -> Tuple[Flask, Datastore]:
+
+    forgetful_node_storage = ForgetfulNodeStorage(federated_only=this_node.federated_only)
 
     from nucypher.characters.lawful import Alice, Ursula
     _alice_class = Alice
@@ -406,7 +422,7 @@ def make_rest_app(
                 raise TemplateError(str(e)) from e
             return Response(response=content, headers=headers)
 
-    return rest_app, datastore
+    return rest_app
 
 
 class TLSHostingPower(KeyPairBasedPower):

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -180,6 +180,7 @@ def check_character_state_after_test(request):
                 offending_tests.add(learner._FOR_TEST)
                 try:  # TODO: Deal with stop vs disenchant.  Currently stop is only for Ursula.
                     learner.stop()
+                    learner._finalize()
                 except AttributeError:
                     learner.disenchant()
             pytest.fail(f"Learners remaining: {still_running}.  Offending tests: {offending_tests} ")

--- a/tests/fixtures.py
+++ b/tests/fixtures.py
@@ -390,6 +390,11 @@ def federated_ursulas(ursula_federated_test_config):
 
     for u in _ursulas:
         u.stop()
+        u._finalize()
+
+    # Pytest will hold on to this object, need to clear it manually.
+    # See https://github.com/pytest-dev/pytest/issues/5642
+    _ursulas.clear()
 
 
 @pytest.fixture(scope="function")
@@ -411,6 +416,8 @@ def lonely_ursula_maker(ursula_federated_test_config):
                 ursula.stop()
             for ursula in self._made:
                 del MOCK_KNOWN_URSULAS_CACHE[ursula.rest_interface.port]
+            for ursula in self._made:
+                ursula._finalize()
     _maker = _PartialUrsulaMaker()
     yield _maker
     _maker.clean()
@@ -697,6 +704,14 @@ def blockchain_ursulas(testerchain, stakers, ursula_decentralized_test_config):
 
     for port in _ports_to_remove:
         del MOCK_KNOWN_URSULAS_CACHE[port]
+
+    for u in _ursulas:
+        u.stop()
+        u._finalize()
+
+    # Pytest will hold on to this object, need to clear it manually.
+    # See https://github.com/pytest-dev/pytest/issues/5642
+    _ursulas.clear()
 
 
 @pytest.fixture(scope="module")

--- a/tests/integration/conftest.py
+++ b/tests/integration/conftest.py
@@ -43,7 +43,6 @@ from tests.constants import (
     MOCK_PROVIDER_URI,
     NUMBER_OF_MOCK_KEYSTORE_ACCOUNTS
 )
-from nucypher.datastore.datastore import Datastore  # yikes TODO
 from tests.fixtures import _make_testerchain, make_token_economics
 from tests.mock.agents import MockContractAgency, MockContractAgent
 from tests.mock.interfaces import MockBlockchain, mock_registry_source_manager
@@ -90,11 +89,6 @@ class TestLMDBEnv:
 @pytest.fixture(autouse=True)
 def JIT_lmdb_env(monkeypatch):
     monkeypatch.setattr("lmdb.open", TestLMDBEnv)
-
-
-@pytest.fixture(autouse=True)
-def reduced_memory_page_lmdb(monkeypatch):
-    monkeypatch.setattr(Datastore, "LMDB_MAP_SIZE", 10_000_000)
 
 
 @pytest.fixture(scope='function', autouse=True)

--- a/tests/integration/network/test_network_upgrade.py
+++ b/tests/integration/network/test_network_upgrade.py
@@ -27,7 +27,7 @@ from nucypher.datastore.models import PolicyArrangement
 from tests.utils.ursula import make_federated_ursulas
 
 
-def test_alice_enacts_policies_in_policy_group_via_rest(enacted_federated_policy, reduced_memory_page_lmdb):
+def test_alice_enacts_policies_in_policy_group_via_rest(enacted_federated_policy):
     """
     Now that Alice has made a PolicyGroup, she can enact its policies, using Ursula's Public Key to encrypt each offer
     and transmitting them via REST.

--- a/tests/mock/performance_mocks.py
+++ b/tests/mock/performance_mocks.py
@@ -195,9 +195,9 @@ class NotARestApp:
 
     def actual_rest_app(self):
         if self._actual_rest_app is None:
-            self._actual_rest_app, _datastore = make_rest_app(db_filepath=tempfile.mkdtemp(),
-                                                             this_node=self.this_node,
-                                                             serving_domains=(None,))
+            self._actual_rest_app, self._datastore = make_rest_app(db_filepath=tempfile.mkdtemp(),
+                                                                   this_node=self.this_node,
+                                                                   serving_domains=(None,))
             _new_view_functions = self._ViewFunctions(self._actual_rest_app.view_functions)
             self._actual_rest_app.view_functions = _new_view_functions
             self._actual_rest_apps.append(

